### PR TITLE
Add SMT-based spec

### PR DESCRIPTION
**Pull request by thpani**
Add a hand-written SMT-based specification of 3SF.

This spec takes a constraint-based approach to encode 3SF and accountable safety.
It uses the decision procedure for [finite sets and cardinality constraints](https://cvc5.github.io/docs-ci/docs-main/theories/sets-and-relations.html) in CVC5.

It was produced as a byproduct of the model-checking work for 3SF, to rule out performance issues caused by Apalache's SMT encoding.

### CODE REVIEW

File: spec-smt/benchmark_ssf.py [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+# This script generates benchmarks with increasing number of blocks and
+# checkpoints, and times CVC5 on each of them. The output is written to
+# stdout in CSV format.
+#
+# WARNING!  This simply replaces a few lines in the model file by offsets.
+# BRITTLE!  If the model file has changed, this script will break!
+
+MODEL_FILENAME = 'ssf.smt'
+BENCHMARK_FILENAME = 'benchmark.smt'
+DATATYPE_LINE_OFFSETS = (28, 32)    # (last line to print, last line to exclude)
```
Comment (Kukovec) [inline spec-smt/benchmark_ssf.py:14]: From the current file, shouldn't this be `(29,33)` ?

File: spec-smt/ssf.smt [hunk 2]
```diff
@@ -0,0 +1,327 @@
...
+(assert
+    (and
+        (exists ((block1 Hash) (block2 Hash))
+            (and
+                (set.member (tuple block1 block2) conflicting_blocks)
+                (set.member block1 finalized_blocks)
+                (set.member block2 finalized_blocks)
+            )
+        )
+        (not (exists ((node Node)) (set.member node slashable_nodes)))
```
Comment (Kukovec) [inline spec-smt/ssf.smt:322]: Would it be cheaper to write this quantifier over finalized checkpoints (and defining `conflicting_checkpoints` analogously) and map the blocks out? You could avoid axiomatizing `finalized_blocks`.

**Review by Kukovec — APPROVED** [review]
LGTM, some minor questions.

File: spec-smt/benchmark_ssf.py (hunk 1 above)
Comment (thpani) [reply spec-smt/benchmark_ssf.py:14]: Line 29 should already be excluded, it declares the number of hashes.
Line 33 should be included, `declare-datatype Vote` is independent of the size parameters.

File: spec-smt/ssf.smt (hunk 2 above)
Comment (thpani) [reply spec-smt/ssf.smt:322]: True, I tried and it's slower for the two smaller benchmark parameters though.
So let's keep it as is.

**Review by konnov — APPROVED** [review]
Very clear encoding! I like it!

**Comment by thpani** [comment]
I initially forgot to add the cardinality check for 1/3 in `AccountableSafety`.
This has been fixed in [96c86f3](https://github.com/freespek/ssf-mc/pull/49/commits/96c86f3befd42055f8319d9d5cf130c81c802096).

@konnov @Kukovec PTAL

File: abstract-spec/smt-spec/ssf.smt [hunk 3]
```diff
@@ -0,0 +1,329 @@
...
+; checkpoint slot
+(declare-fun checkpoint_slot (Checkpoint) Int)
+(assert (= (checkpoint_slot genesis_checkpoint) 0))
+
+; all checkpoints ((B, block_slot), chkp_slot) except genesis have chkp_slot > block_slot
+(assert (forall ((c Checkpoint)) (or (= c genesis_checkpoint) (> (checkpoint_slot c) (slot (checkpoint_block c))))))
+
+; checkpoints have no "identity": if they have the same block and slot, they are the same checkpoint
+(assert (forall ((c1 Checkpoint) (c2 Checkpoint))
+    (=>
```
Comment (konnov) [inline abstract-spec/smt-spec/ssf.smt:100]: should not that be `=`, that is, "if and only if"?

File: abstract-spec/smt-spec/ssf.smt [hunk 4]
```diff
@@ -0,0 +1,329 @@
...
+    )
+))
+
+; ============================
+; Votes
+; ============================
+
+; votes have no "identity": if they have the same checkpoints and sender, they are the same vote
+(assert (forall ((v1 Vote) (v2 Vote))
+    (=>
```
Comment (konnov) [inline abstract-spec/smt-spec/ssf.smt:115]: should not that be `=`, that is, "if and only if"?

**Review by konnov — APPROVED** [review]
This is amazing. Just left a few comments.

File: abstract-spec/smt-spec/ssf.smt (hunk 3 above)
Comment (thpani) [reply abstract-spec/smt-spec/ssf.smt:100]: It's not wrong, but the `<=` direction should be implied by EUF? (`checkpoint_block`/`checkpoint_slot` are functions)

File: abstract-spec/smt-spec/ssf.smt (hunk 4 above)
Comment (thpani) [reply abstract-spec/smt-spec/ssf.smt:115]: Actually, both directions are implied by `Vote` being a datatype (DT instances are equal iff they are structurally equal).

It looks like I kept this constraint from a previous version where `source`/`target`/`sender` where functions over `Vote`s, but I don't think it affects solving time a lot.

File: abstract-spec/smt-spec/benchmark_ssf.py [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+    result = subprocess.run(['cvc5', BENCHMARK_FILENAME], capture_output=True)
+    end = time.time()
+
+    time = end - start
+    output = result.stdout.decode().strip().split('\n')[-1]
+    return (time, output)
+
+print(f'blocks,checkpoints,result,time')
+for checkpoints in range(5, 8):
+    for blocks in range (3, 6):
```
[the remark this answers is not in the document]
Comment (thpani) [reply abstract-spec/smt-spec/benchmark_ssf.py]: good point, done!

File: abstract-spec/smt-spec/benchmark_ssf.py [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+# stdout in CSV format.
+#
+# Thomas Pani, 2024
+#
+# WARNING!  This simply replaces a few lines in the model file by offsets.
+# BRITTLE!  If the model file has changed, this script will break!
+
+MODEL_FILENAME = 'ssf.smt'
+BENCHMARK_FILENAME = 'benchmark.smt'
+DATATYPE_LINE_OFFSETS = (29, 34)    # (last line to print, last line to exclude)
```
[the remark this answers is not in the document]
Comment (thpani) [reply abstract-spec/smt-spec/benchmark_ssf.py:19]: Well, maybe, but I don't expect the spec to change anytime soon.

If it does, it's very likely that this script has to be revisited anyhow.
Also, CVC5 will fail if you get the offsets wrong here, so I don't think it matters much.

**Pull request merged**